File: python/parse.py
from enum import Enum
from dataclasses import dataclass
from typing import List, Union

class ExprType(Enum):
    VAR = 1
    INT_LIT = 2
    STRING_LIT = 3
    APP = 4
    LAMBDA = 5
    LET = 6

@dataclass
class Expr:
    type: ExprType
    value: Union[str, int, None] = None
    children: List['Expr'] = None

@dataclass
class Decl:
    name: str
    params: List[str]
    body: Expr

@dataclass
class Script:
    declarations: List[Decl]

class Parser:
    def __init__(self, input: str):
        self.input = input
        self.pos = 0

    def parse(self) -> Script:
        script = Script(declarations=[])
        while self.pos < len(self.input):
            script.declarations.append(self.parse_decl())
        return script

    def parse_decl(self) -> Decl:
        name = self.parse_identifier()
        params = []
        while self.peek() != '=':
            params.append(self.parse_identifier())
        self.consume('=')
        body = self.parse_expr()
        self.consume(';')
        return Decl(name, params, body)

    def parse_expr(self) -> Expr:
        if self.peek().isalpha():
            return Expr(ExprType.VAR, value=self.parse_identifier())
        raise ValueError("Unsupported expression")

    def parse_identifier(self) -> str:
        ident = ""
        while self.pos < len(self.input) and self.input[self.pos].isalpha():
            ident += self.input[self.pos]
            self.pos += 1
        if not ident:
            raise ValueError("Expected identifier")
        return ident

    def peek(self) -> str:
        return self.input[self.pos] if self.pos < len(self.input) else ''

    def consume(self, expected: str):
        if self.peek() != expected:
            raise ValueError(f"Expected '{expected}', got '{self.peek()}'")
        self.pos += 1

File: go/parse.go
package miranda

import (
    "fmt"
    "strings"
    "unicode"
)

type ExprType int

const (
    VarExpr ExprType = iota
    IntLitExpr
    StringLitExpr
    AppExpr
    LambdaExpr
    LetExpr
)

type Expr struct {
    Type     ExprType
    Value    interface{}
    Children []*Expr
}

type Decl struct {
    Name   string
    Params []string
    Body   *Expr
}

type Script struct {
    Declarations []*Decl
}

type Parser struct {
    input string
    pos   int
}

func NewParser(input string) *Parser {
    return &Parser{input: input, pos: 0}
}

func (p *Parser) parse() (*Script, error) {
    script := &Script{}
    for p.pos < len(p.input) {
        decl, err := p.parseDecl()
        if err != nil {
            return nil, err
        }
        script.Declarations = append(script.Declarations, decl)
    }
    return script, nil
}

func (p *Parser) parseDecl() (*Decl, error) {
    name, err := p.parseIdentifier()
    if err != nil {
        return nil, err
    }
    params := []string{}
    for p.peek() != '=' {
        param, err := p.parseIdentifier()
        if err != nil {
            return nil, err
        }
        params = append(params, param)
    }
    p.consume('=')
    body, err := p.parseExpr()
    if err != nil {
        return nil, err
    }
    p.consume(';')
    return &Decl{Name: name, Params: params, Body: body}, nil
}

func (p *Parser) parseExpr() (*Expr, error) {
    // Simplified expression parsing
    if unicode.IsLetter(rune(p.peek())) {
        ident, err := p.parseIdentifier()
        if err != nil {
            return nil, err
        }
        return &Expr{Type: VarExpr, Value: ident}, nil
    }
    return nil, fmt.Errorf("unsupported expression")
}

func (p *Parser) parseIdentifier() (string, error) {
    start := p.pos
    for p.pos < len(p.input) && unicode.IsLetter(rune(p.input[p.pos])) {
        p.pos++
    }
    if start == p.pos {
        return "", fmt.Errorf("expected identifier")
    }
    return p.input[start:p.pos], nil
}

func (p *Parser) peek() byte {
    if p.pos >= len(p.input) {
        return 0
    }
    return p.input[p.pos]
}

func (p *Parser) consume(expected byte) error {
    if p.peek() != expected {
        return fmt.Errorf("expected '%c', got '%c'", expected, p.peek())
    }
    p.pos++
    return nil
}